perfkitbenchmarker/background_tasks.py

```diff
@@ -0,0 +1,633 @@
+  target, args, kwargs = target_arg_tuple
+  while isinstance(target, functools.partial):
+    args = target.args + args
+    inner_kwargs = target.keywords.copy()
+    inner_kwargs.update(kwargs)
+    kwargs = inner_kwargs
+    target = target.func
+  arg_strings = [str(a) for a in args]
+  arg_strings.extend(['{0}={1}'.format(k, v) for k, v in kwargs.iteritems()])
+  return '{0}({1})'.format(getattr(target, '__name__', target),
+                           ', '.join(arg_strings))
+
+
+def _WaitForCondition(condition_callback, timeout=None):
+  """Waits until the specified callback returns a value that evaluates True.
+
+  Similar to the threading.Condition.wait method that is the basis of most
+  threading class wait routines. Polls the condition, starting with frequent
+  checks but extending the delay between checks upon each failure.
+
+  Args:
+    condition_callback: Callable that returns a value that evaluates True to end
+        the wait or evaluates False to continue the wait.
+    timeout: Optional float. Number of seconds to wait before giving up. If
+        provided, the condition is still checked at least once before giving up.
+        If not provided, the wait does not time out.
+
+  Returns:
+    True if condition_callback returned a value that evaluated True. False if
+    condition_callback did not return a value that evaluated True before the
+    timeout.
+  """
+  deadline = None if timeout is None else time.time() + timeout
+  delay = _WAIT_MIN_RECHECK_DELAY
+  while True:
+    if condition_callback():
+      return True
+    remaining_time = (_WAIT_MAX_RECHECK_DELAY if deadline is None
+                      else deadline - time.time())
+    if remaining_time <= 0:
+      return False
+    time.sleep(delay)
+    delay = min(delay * 2, remaining_time, _WAIT_MAX_RECHECK_DELAY)
+
+
+class _SingleReaderQueue(object):
+  """Queue to which multiple threads write but from which only one thread reads.
+
+  A lightweight substitute for the Queue.Queue class that does not use
+  internal Locks.
+
+  Gets are interruptable but depend on polling.
+  """
+
+  def __init__(self):
+    self._deque = deque()
+
+  def Get(self, timeout=None):
+    if not _WaitForCondition(lambda: self._deque, timeout):
+      raise Queue.Empty
+    return self._deque.popleft()
+
+  def Put(self, item):
+    self._deque.append(item)
+
+
+class _NonPollingSingleReaderQueue(object):
+  """Queue to which multiple threads write but from which only one thread reads.
+
+  Uses a threading.Lock to implement a non-interruptable Get that does not poll
+  and is therefore easier on CPU usage. The reader waits for items by acquiring
+  the Lock, and writers release the Lock to signal that items have been written.
+  """
+
+  def __init__(self):
+    self._deque = deque()
+    self._lock = threading.Lock()
+    self._lock.acquire()
+
+  def _WaitForItem(self):
+    self._lock.acquire()
+
+  def _SignalAvailableItem(self):
```

Comment:
Seems like [condition variables](https://docs.python.org/2/library/threading.html#condition-objects) might be appropriate here.